util/chunk/list.go

```diff
@@ -16,17 +16,19 @@ package chunk
 import (
 	"github.com/juju/errors"
 	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/memory"
 )
 
 // List holds a slice of chunks, use to append rows with max chunk size properly handled.
 type List struct {
-	fieldTypes       []*types.FieldType
-	maxChunkSize     int
-	length           int
-	chunks           []*Chunk
-	freelist         []*Chunk
-	memUsageChunks   int64
-	memUsageFreelist int64
+	fieldTypes   []*types.FieldType
+	maxChunkSize int
+	length       int
+	chunks       []*Chunk
+	freelist     []*Chunk
+
+	memTracker   *memory.Tracker // track memory usage.
+	lastConsumed bool            // whether the memory usage of last Chunk in "chunks" has been consumed.
```

Comment:
How about using an int variable `unconsumedBytes` that represents unconsumed memory usage?

Reply:
if we track the unconsumed bytes, every time a `List.AppendRow()` function call will cause a `Chunk.MemUsage()` function call, makes it inefficient.

Reply:
No, we can only consume it when adding a new chunk and reset it to 0, otherwise, update `unconsumedBytes`.